### ISSUE TITLE
After edit, Editor opens in wrong position

### ISSUE THREAD

**Issue by alukach** [emory-lits-labs/annotator-imgselect#3]
I've noticed that after a user edits an annotation, any following annotation creation opens the Editor above the annotation that was previously edited.

![anim](https://cloud.githubusercontent.com/assets/897290/17629028/8c2930de-6075-11e6-9fd0-e08cf131f0ef.gif)

I'm going to try digging into this today but I'm curious if @rlskoeser knows anything about why this may be (or can direct me to the tooling that controls where the editor opens).

**Comment by alukach** [comment]
Okay, so going deeper:

I see that in `imgselect_utils.selectionEnd()`, the last step is to set the position of the `annotator-editor` element:

``` js
      // set editor window is not positioned relative to the adder element
      var offset = $(s.adder.element[0]).offset();
      if (offset){
        $(".annotator-editor").css({
          top: offset.top + 50,
          left: offset.left - (selection.width/2)
        });
      }
```

The problem seems to come from the fact that after editing an annotation, the `.css()` call fails to actually update the position of the `.annotator-editor` element in the DOM:

![image](https://cloud.githubusercontent.com/assets/897290/17631345/ff2afb52-6080-11e6-9744-2087048e08af.png)

Running `.attr('style')` on the element reports where the editor _should_ be placed, but examining the element directly shows that the `style` attribute is still set to be where it was when the edit took place (ie in the wrong position).  It's a very strange situation, not sure why the DOM element is failing to be updated when JQuery reports that it's being updated.

**Comment by rlskoeser** [comment]
@alukach I think the editor position is something we've struggled with in some of the other positions too.  I know we're calling the annotation [adder.load](https://github.com/openannotation/annotator/blob/5ef5edf157fe728b2d6d95d01e26a55c508c0c44/src/ui/adder.js#L56) method, I wonder if we should try using the [adder.show](https://github.com/openannotation/annotator/blob/5ef5edf157fe728b2d6d95d01e26a55c508c0c44/src/ui/adder.js#L73) method instead of the custom offset code you've identified.

**Comment by alukach** [comment]
@rlskoeser If I understand correctly, the `adder` is just the little popup helper to enter the editor:

![image](https://cloud.githubusercontent.com/assets/897290/17632073/c71ca802-6083-11e6-9214-78a05a92e0f7.png)

I started off blaming the Adder, however the Adder is always positioned correctly.

Some more notes as I'm looking at this:  If you create and edit a first annotation, and then create a second selection box, the (hidden) Editor is correctly positioned by the `.css()` code written above (and both the DOM and the `.attr` results are in agreement).  It's only after you click the Adder button on the second annotation does the Editor box move back to its incorrect location.

![image](https://cloud.githubusercontent.com/assets/897290/17632205/5dd83018-6084-11e6-9e1e-9096a7ce220f.png)

**Comment by rlskoeser** [comment]
You're absolutely right about the adder. I guess it's been a while since I worked on this!  What we want is the editor, then, not the adder.  But it looks like we don't call `editor.load` method directly, I guess the adder handles that....  So I wonder if the css offset is getting applied to the wrong element, or happening too soon?  It seems like it is getting ignored entirely, since it's in the same position as before.

I think maybe this is [where the editor gets loaded](https://github.com/openannotation/annotator/blob/5ef5edf157fe728b2d6d95d01e26a55c508c0c44/src/ui/main.js#L304) - and it looks the main.ui [text selector updates that interaction point](https://github.com/openannotation/annotator/blob/5ef5edf157fe728b2d6d95d01e26a55c508c0c44/src/ui/main.js#L249) - wonder if we could update it too?

**Comment by alukach** [comment]
You're spot-on in your assessment.

Okay, I've set a DOM Breakpoint (something I never knew about until today) on the `.annotator-editor` element.  I can now see that when you click on the Adder, it fires a Click event that is caught be `adder._onClick`, which fires our `onCreate` method, which runs `app.annotations.create(ann);`, which eventually gets to [`beforeAnnotationCreated()`](https://github.com/openannotation/annotator/blob/5ef5edf157fe728b2d6d95d01e26a55c508c0c44/src/ui/main.js#L299-L305) (as you said) which returns `s.editor.load(annotation, s.interactionPoint);`, but `s.interactionPoint` is incorrect. The editor is then loaded at the position of the incorrect `InteractionPoint`.

So, the question is _how_ we can update `s.interactionPoint`.  It's confusing to me because `s` is a variable local to the `main()` function:

``` js
function main(options) {
    if (typeof options === 'undefined' || options === null) {
        options = {};
    }

    options.element = options.element || global.document.body;
    options.editorExtensions = options.editorExtensions || [];
    options.viewerExtensions = options.viewerExtensions || [];

    // Local helpers
    var makeAnnotation = annotationFactory(options.element, '.annotator-hl');

    // Object to hold local state
    var s = {
        interactionPoint: null
    };
```

The TextSelector can access/update that variable because its function is defined within `main()`, thus making `s` available via lexical scoping. I don't know if there will be any way for us to access that variable being that we are defining our tooling outside of the `main()` function.

**Comment by rlskoeser** [comment]
Wow, you're right.  Thanks for opening the issue on annotator.js, maybe we'll get some useful feedback.  I know there are some issues about adding some hooks to make things more available to plugins, maybe there's something that's needed here so we can do things better.  It is entirely possible that the plugin isn't doing the right thing as you said in your ticket - to some degree I felt like I was working around annotator more than I would have liked.

I do wonder if there is a way to override/replace the [main annotation hooks](https://github.com/openannotation/annotator/blob/5ef5edf157fe728b2d6d95d01e26a55c508c0c44/src/ui/main.js#L299-L310) that display the editor - but if so, then we'd need to do it in a way that it only affects the image selection annotations and not the text selections.
